sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,502 @@
+  is submitted to a device that does not support those features.  This
+  exception must be raised synchronously from the kernel invocation command
+  (e.g. `parallel_for()`).
+
+* The runtime must not raise an exception (or otherwise fail) merely due to
+  speculative compilation of a kernel for a device, when the application does
+  not specifically submit the kernel to that device.
+
+### Diagnostics from the front-end compiler
+
+By "front-end compiler", we mean the DPC++ compiler which parses DPC++ source
+code, not the JIT compiler that translates SPIR-V into native code.
+
+In general, the front-end compiler does not know which kernels the application
+will submit to which devices.  Therefore, the front-end compiler does not
+generally know which optional features a kernel can legally use.  Thus, in
+general, the front-end compiler must not issue any diagnostic simply because a
+kernel uses an optional feature.
+
+The only exception to this rule occurs when the application uses the C++
+attribute `[[sycl::requires()]]`.  When the application decorates a kernel or
+device function with this attribute, it is an assertion that the kernel or
+device function is expected to use *only* the features listed by the attribute.
+Therefore, the front-end compiler must issue a diagnostic if the kernel or
+device function uses any other optional kernel features.
+
+Note that this behavior does not change when the compiler runs in AOT mode.
+Even if the user specifies a target device via "-fsycl-targets", that does not
+necessarily mean that the user expects all the code in the application to be
+runnable on that device.  Consider an application that uses some middleware
+library, where the library's header contains kernels optimized for different
+devices.  An application should be able to compile in AOT mode with this
+library without getting errors.  Therefore the AOT compiler must not fail
+simply because the middleware header contains device code for devices that
+are not being compiled for.
+
+### Runtime exception if device doesn't support feature
+
+When the application submits a kernel to a device via one of the the kernel
+invocation commands (e.g. `parallel_for()`), the runtime must check to see
+if the kernel uses optional features that are not supported on that device.
+If the kernel uses a feature that is not supported, the runtime must throw
+a synchronous `errc::kernel_not_supported` exception.
+
+This exception must be thrown in the following circumstances:
+
+* A device function in the kernel's static call tree uses a feature that the
+  device does not support.  However, this only applies to features that are
+  exposed via a C++ type or function.  Examples of this include `sycl::half` or
+  instantiating `sycl::atomic_ref` for a 64-bit type.  In cases where the
+  feature is more "notional", such as requiring a particular type of forward
+  progress guarantee, no exception is required.
```

Comment:
Could a more particular example of an optional forward progress guarantee be provided here? Or is it better for users to rely on the SYCL-2020 spec itself?

Reply:
I reworded to reference the sub-group forward progress guarantee in 333bf432e75db.